File: problems/avoid-puddles.py
# https://programmers.co.kr/learn/courses/30/lessons/42898

def solution(width, height, obstacles):
    markedMap = [[0 for i in range(width)] for j in range(height)]
    markedMap[0][0] = 1
    for obstacle in obstacles:
        markedMap[obstacle[0]-1][obstacle[1]-1] = -1
    for i in range(height):
        for j in range(width):
            if markedMap[i][j] != -1:
                if i - 1 >= 0 and markedMap[i-1][j] != -1:
                    markedMap[i][j] += markedMap[i-1][j] % 1000000007
                if j - 1 >= 0 and markedMap[i][j-1] != -1:
                    markedMap[i][j] += markedMap[i][j-1] % 1000000007
    return markedMap[-1][-1]

print(solution(4,3,[[2,2]]))

File: problems/programmers.co.kr/avoid-puddles.js
function solution(puddles, width, height){
	let startX = 0
	let startY = 0
	const row = [...Array(width)].map(elem=>0)
	let markedMap = [...Array(height)].map(elem=>[...row])
	markedMap[0][0] = 1
	obstacles.forEach(obstacle => {
		markedMap[obstacle[0]][obstacle[1]] = -1
    })
	for(let i = 0; i < width; i++){
		for(let j = 0; j < height; j++){
			if(markedMap[i][j] !== -1){ 
                if(i - 1 >= 0 && markedMap[i-1][j] !== -1)
                    markedMap[i][j] += markedMap[i-1][j]
                if(j - 1 >= 0 && markedMap[i][j-1] !== -1) 
                    markedMap[i][j] += markedMap[i][j-1]
            }
        }
    }
	return markedMap
}
